### PR TITLE
Add profitability module

### PR DESCRIPTION
**Pull request by chrismaree**
Adds profitability module to relayer to consider relay profit before executing.

### CODE REVIEW

File: src/clients/ProfitClient.ts [hunk 1]
```diff
@@ -0,0 +1,122 @@
+import { BigNumber, winston, toBNWei, toBN, assign } from "../utils";
+import { HubPoolClient } from ".";
+import { Deposit, L1Token } from "../interfaces";
+import { Coingecko } from "@uma/sdk";
+
+// Define the minimum revenue, in USD, that a relay must yield in order to be considered "profitable". This is a short
+// term solution to enable us to avoid DOS relays that yield negative profits. In the future this should be updated
+// to actual factor in the cost of sending transactions on the associated target chains.
+const chainIdToMinRevenue = {
```
Comment (chrismaree) [inline src/clients/ProfitClient.ts:9]: this is the temp workaround we've spoken about before. at minimum each of these chains must produce the amount of USD revenue defined in this const object.

File: src/relayer/Relayer.ts [hunk 2]
```diff
@@ -99,8 +108,35 @@ export class Relayer {
...
+          `${getNetworkName(deposit.originChainId)} to ${getNetworkName(deposit.destinationChainId)}` +
+          ` and an unfilled amount of  ${formatFunction(fillAmount)} ${symbol} is unprofitable!\n`;
+      });
+    });
+
+    this.logger.warn({ at: "Relayer", message: "Not relaying unprofitable deposits 🙅‍♂️!", mrkdwn });
+
+    this.clients.profitClient.clearUnprofitableFills();
 
+    // Note that this implementation right now will result in each run of the bot logging for insufficient token balance.
```
Comment (chrismaree) [inline src/relayer/Relayer.ts:138]: this note is important.

File: src/relayer/Relayer.ts [hunk 3]
```diff
@@ -81,12 +87,15 @@ export class Relayer {
     return unfilledDeposits;
   }
 
+  // TODO: that the implementations below for both methods will produce logs on each iteration of the bot. This should
```
Comment (chrismaree) [inline src/relayer/Relayer.ts:90]: this note is important. we can fix both of these in a subsequent PR.

File: src/clients/ProfitClient.ts [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+  }
+
+  getAllPrices() {
+    return this.tokenPrices;
+  }
+
+  getPriceOfToken(token: string) {
+    if (!this.tokenPrices[token]) {
+      this.logger.warn({ at: "ProfitClient", message: `Token ${token} not found in state. Using 0` });
+      return toBN(0);
```
Comment (mrice32) [inline src/clients/ProfitClient.ts:45]: Just double checking my understanding, a price of 0 means we will _not_ relay this token, right?

File: src/clients/ProfitClient.ts [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+  isFillProfitable(deposit: Deposit, fillAmount: BigNumber) {
+    if (toBN(this.relayerDiscount).eq(toBNWei(1))) {
+      this.logger.debug({ at: "ProfitClient", message: "Relayer discount set to 100%. Accepting relay" });
+      return true;
+    }
+
+    if (toBN(deposit.relayerFeePct).eq(toBN(0))) {
+      this.logger.debug({ at: "ProfitClient", message: "Deposit set 0 relayerFeePct. Rejecting relay" });
+      return false;
+    }
```
[the remark this answers is not in the document]
Comment (chrismaree) [reply src/clients/ProfitClient.ts:67]: yup! that's why I added it :)

File: src/clients/ProfitClient.ts [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+    const { decimals, address: l1Token } = this.hubPoolClient.getTokenInfoForDeposit(deposit);
+    const tokenPriceInUsd = this.getPriceOfToken(l1Token);
+    const fillRevenueInRelayedToken = toBN(deposit.relayerFeePct).mul(fillAmount).div(toBN(10).pow(decimals));
+    const fillRevenueInUsd = fillRevenueInRelayedToken.mul(tokenPriceInUsd).div(toBNWei(1));
+    const discount = toBNWei(1).sub(this.relayerDiscount);
```
[the remark this answers is not in the document]
Comment (chrismaree) [reply src/clients/ProfitClient.ts:72]: not quite. consider it in the following line:
```
const discount = toBNWei(1).sub(this.relayerDiscount);
    const minimumAcceptableRevenue = chainIdToMinRevenue[deposit.destinationChainId].mul(discount).div(toBNWei(1));
```
what I've called "discount" is actually how much the `chainIdToMinRevenue` is scaled _down_ by, indicating the minimum acceptable revenue. I think the variable name `discount` is what is making this confusing.

I've renamed this to `revenueScalar` and added a comment to be more clear.

File: src/clients/ProfitClient.ts (hunk 1 above)
Comment (chrismaree) [reply src/clients/ProfitClient.ts:9]: we can always increase it. my thinking with 10 USD is it is above what you'd do to spam the bot but is ~ what a deposit is likely to cost. (an erc20 transfer at 30 gwei is like 7 USD)

**Pull request merged**